libs/pika/async_mpi/include/pika/async_mpi/trigger_mpi.hpp

```diff
@@ -0,0 +1,263 @@
+
+/// \file parallel/algorithms/transform_xxx.hpp
+
+#pragma once
+
+#include <pika/config.hpp>
+#include <pika/assert.hpp>
+#include <pika/async_mpi/dispatch_mpi.hpp>
+#include <pika/async_mpi/mpi_exception.hpp>
+#include <pika/async_mpi/mpi_polling.hpp>
+#include <pika/concepts/concepts.hpp>
+#include <pika/datastructures/variant.hpp>
+#include <pika/debugging/demangle_helper.hpp>
+#include <pika/debugging/print.hpp>
+#include <pika/execution/algorithms/detail/helpers.hpp>
+#include <pika/execution/algorithms/detail/partial_algorithm.hpp>
+#include <pika/execution/algorithms/transfer.hpp>
+#include <pika/execution_base/any_sender.hpp>
+#include <pika/execution_base/receiver.hpp>
+#include <pika/execution_base/sender.hpp>
+#include <pika/executors/thread_pool_scheduler.hpp>
+#include <pika/functional/detail/tag_fallback_invoke.hpp>
+#include <pika/functional/invoke.hpp>
+#include <pika/functional/invoke_fused.hpp>
+#include <pika/mpi_base/mpi.hpp>
+
+#include <exception>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace pika::mpi::experimental::detail {
+
+    namespace pud = pika::util::detail;
+    namespace ex = pika::execution::experimental;
+
+    // -----------------------------------------------------------------
+    // route calls through an impl layer for ADL isolation
+    template <typename Sender>
+    struct trigger_mpi_sender_impl
+    {
+        struct trigger_mpi_sender_type;
+    };
+
+    template <typename Sender>
+    using trigger_mpi_sender = typename trigger_mpi_sender_impl<Sender>::trigger_mpi_sender_type;
+
+    // -----------------------------------------------------------------
+    // transform MPI adapter - sender type
+    template <typename Sender>
+    struct trigger_mpi_sender_impl<Sender>::trigger_mpi_sender_type
+    {
+        using is_sender = void;
+        std::decay_t<Sender> sender;
+        int completion_mode_flags_;
+
+#if defined(PIKA_HAVE_STDEXEC)
+        using completion_signatures = pika::execution::experimental::completion_signatures<
+            pika::execution::experimental::set_value_t()>;
+#else
+        // -----------------------------------------------------------------
+        // completion signatures
+        template <template <typename...> class Tuple, template <typename...> class Variant>
+        using value_types = Variant<Tuple<>>;
+
+        template <template <typename...> class Variant>
+        using error_types = pud::unique_t<pud::prepend_t<
+            typename ex::sender_traits<Sender>::template error_types<Variant>, std::exception_ptr>>;
+#endif
+
+        static constexpr bool sends_done = false;
+
+        // -----------------------------------------------------------------
+        // operation state for an internal receiver
+        template <typename Receiver>
+        struct operation_state
+        {
+            std::decay_t<Receiver> receiver;
+            int mode_flags_;
+            int status;
```

Comment:
Is `status` always initialized? Codacy doesn't seem to think so but it may be wrong.

Reply:
no it isn't, but it is only used by one of the modes (suspend/resume), the others call set_error when necessary without needing to store it. (default initialization would not hurt, but ...)